programs/storage_api/src/storage_contract.rs

```diff
@@ -447,24 +470,48 @@ fn store_validation_result(
                 return Err(InstructionError::InvalidAccountData);
             }
 
-            reward_validations
-                .entry(segment)
-                .or_default()
-                .insert(*me, proof_mask.to_vec());
+            let (recorded_validations, _) = count_valid_proofs(&validations);
+            let entry = validations.entry(segment).or_default();
+            if !entry.contains_key(me) {
+                entry.insert(*me, proof_mask.to_vec());
+            }
+            let (total_validations, _) = count_valid_proofs(&validations);
+            update_credits(credits, current.epoch);
+            credits.current_epoch += total_validations - recorded_validations;
         }
         _ => return Err(InstructionError::InvalidAccountData),
     }
     storage_account.account.set_state(&storage_contract)
 }
 
-fn count_valid_proofs(proofs: &[ProofStatus]) -> u64 {
+fn update_credits(credits: &mut Credits, current_epoch: u64) {
```

Comment:
this function looks like it would tally credits, but it's really just epoch-shifting?

Reply:
you didn't want to put this method on Credits?

Reply:
You're right, better suited to live in Credits, update_epoch probably?